docs/overview.md

```diff
@@ -0,0 +1,670 @@
+## Overview
+
+<!--
+The source of this file contains inline HTML comments starting with '[x]'.
+These are the lines from the test results in testdox format, which are verifying the statement made
+in this documentation.
+
+Use this as an example on how to turn test documentation into user documentation.
+Of course that requires a good testbase.
+
+To reproduce the test results, run `phpunit --testdox`
+-->
+The Joomla! Dependency Injection package provides a simple `container-interop` (upcoming `PSR-11`) compatible
+Inversion of Control (IoC) Container for your application.
+
+The Joomla! Dependency Injection Container supports
+
+  - **factories** and **instances** (i.e., **closures**, **callables**, **objects**, **arrays**, and even **scalars**)
+  - **shared** and **protected** modes
+  - **caching** and **cloning** of resources according to the modes
+  - enforcing **recreation** of resources
+  - **aliases**
+  - **scopes** using the **decorator** pattern for containers
+  - **delegate lookup**
+  - **object creation** on the fly to resolve dependencies
+  - ability to **extend** resources
+  - **service providers**
+
+In this document,
+
+  - **factory** is a callable or closure, that takes the container as argument and returns the resource instance;
+  - **instance** is the resource instance, i.e., the return value of a factory, or an explicitly defined value (may be a scalar as well);
+  - **resource** is a key/value pair, with the key being the id, and the value being a factory or an instance.
+
+### Container Interoperability
+    
+The Joomla! Dependency Injection package implements the [PSR-11 proposal](https://github.com/container-interop/fig-standards/blob/master/proposed/container.md)
+for Dependency Injection Containers to achieve interoperability.
+Until PSR-11 gets accepted, Joomla! DI uses the [`container-interop`](https://github.com/container-interop/container-interop)
+namespace.
+
+### Creating a Container object
+
+Creating a container usually happens very early in the application lifecycle. For a Joomla! MVC app, this
+typically happens in the application's `doExecute` method. This allows your application access to the DI
```

Comment:
I don't know if there's a right or wrong here and DI's something that will be debated until PHP dies.  The approach I've taken with newer Framework apps is that the container is instantiated with all of the application service providers (ironically including a provider for the application class) before instantiating the application since we do not have a compiled container (compare to Symfony for this one).  So once the application is booted, the fully compiled container is injected into it as a dependency of the application versus having the application assemble it.  That feels more "correct" to me personally, but YMMV.